### ISSUE TITLE
The "cause" property of Error is ignored when matching

### ISSUE THREAD

**Issue by kettanaito** [vitest-dev/vitest#5697]
### Describe the bug

Vitest ignores the `cause` property whenever performing any equality on `Error` instances.

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-fq9xyu?file=test%2Fbasic.test.ts&initialPath=__vitest__/

### System Info

```shell
Irrelevant. The issue is in the matcher.
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
This is documented behavior: https://vitest.dev/api/expect.html#toequal

> A deep equality will not be performed for Error objects. Only the message property of an Error is considered for equality. To customize equality to check properties other than message, use [expect.addEqualityTesters](https://vitest.dev/api/expect.html#expect-addequalitytesters). To test if something was thrown, use [toThrowError](https://vitest.dev/api/expect.html#tothrowerror) assertion.

**Comment by kettanaito** [comment]
Hi, @sheremet-va. Got it. Can you give me an example of using `.toThrowError()` to also assert the `cause` property, please? I find it to be a common use case, I wouldn't want to add a custom tester/matcher for it.

**Comment by kettanaito** [comment]
I don't believe `.toThrowError()` allows me to achieve the goal here. I highly suggest Vitest adds an API to help test more complex errors easier. 

Perhaps something like this?

```ts
const error = vi.throws(fn)
expect(error).toEqual({ message: '', cause: '' })
```

**Comment by kettanaito** [comment]
@sheremet-va, also, any chance to revisit this?

> Only the message property of an Error is considered for equality.

I assume `message` is checked because every error can have a message. The same way, every error can have a `cause` property.

**Comment by sheremet-va** [comment]
Technically you can use chai API for this:

```ts
expect(() => {
  throw new Error('test', { cause: new Error('cause') });
})
  .throws()
  .property('cause')
  .eql(new Error('cause'))
```

I am not sure what the best course of action is here, I also don't really like that it only checks the message. Will put it on the board to discuss with the team. If anyone has any ideas here, feel free to suggest here.

Maybe adding `toThrowStrictError` matcher? 🤔

**Comment by kettanaito** [comment]
I think there are two possible directions to this:

1. Add `cause` as a special case alongside `message`, which equality will always be checked.
2. Add `.toThrowStrictError()`, although why then differentiate between this and `.toThrowError`? To keep the previous behavior intact? Is the intention behind `.toThrowStrictError()` to strictly compare _all_ error properties? 

> Technically you can use chai API for this:

That's a good call! I will remember this for myself but it won't work for what I'm preparing for my students right now. 

Not checking arbitrary properties is okay. But `cause` is making its way into our code, and it's a defined property with a clear purpose (and also value type). I'd expect this to work:

```ts
expect(fn).toThrow(new Error('message', { cause: 123 }))
```

> Need to be careful since `cause` is a potentially deep property. It can reference an error which has its own `cause`, which points to an error, which...

**Comment by sheremet-va** [comment]
I see your point. It would be nice to check `cause` since it's in the standard now. I believe we should also check the cause's cause and so on if it wasn't validated yet (hello, recursion).

Not sure about `vi.throws` - we have a few non-testing utilities already, so there is nothing stopping us from adding it.

**Comment by kettanaito** [comment]
> I believe we should also check the cause's cause and so on if it wasn't validated yet 

Only if explicitly specified in the assertion 👍 

```js
const error = new Error('message', {
  cause: new Error('another', {
    cause: 123
  })
})

expect(error).toEqual(new Error('message')) // OK!
expect(error).toEqual(new Error('message', { cause: new Error('another') })) //  OK!
expect(error).toEqual(new Error('message', { cause: new Error('another', { cause: 123 }) })) //  OK!

expect(error).toStrictEqual(new Error('message')) // X
expect(error).toStrictEqual(new Error('message', { cause: new Error('another') })) // X
expect(error).toStrictEqual(new Error('message', { cause: new Error('another', { cause: 'bad' }) })) //  X
expect(error).toEqual(new Error('message', { cause: new Error('another', { cause: 'bad' }) })) //  X
```

**Comment by hi-ogawa** [comment]
Something similar is brought up here
- https://github.com/vitest-dev/vitest/issues/5244

Comparing only `Error.message` is probably a historical reason and we should probably revisit this. In the issue, I was mentioning how `node:assert`'s deep equality does https://github.com/vitest-dev/vitest/issues/5244#issuecomment-1953886008 since they treat `Error` differently.

It turns out Node doesn't check `Error.cause` since it's not enumerable. They only treat `Error.name` and `Error.message` as a special non-enumerable property to check.
https://stackblitz.com/edit/vitest-dev-vitest-ma1ej3?file=test%2Fbasic.test.ts

```ts
import assert from "node:assert";

// ok
assert.deepStrictEqual(
  new Error('x', { cause: 'foo' }),
  new Error('x', { cause: 'bar' })
);
```

This is not stopping Vitest from checking `Error.cause` but I thought it's worth mentioning.

**Comment by hi-ogawa** [comment]
Update from my last comment, Node v22 actually started to check `Error.cause` and `AggregateError.errros` for the equality.

https://nodejs.org/docs/latest-v22.x/api/assert.html#assertdeepstrictequalactual-expected-message

> Error names, messages, causes, and errors are always compared, even if these are not enumerable properties. errors is also compared.

I made a quick repro here https://github.com/hi-ogawa/reproductions/tree/main/node-error-equality and I can verify the new behavior, but they probably haven't considered adjusting error diff, so their current assertion error message is quite cryptic as it doesn't include `cause` diff.

It looks like this is added quite recently (and maybe prematurely?) https://github.com/nodejs/node/pull/51805, so we might want to wait a bit longer how it turns out in node world.

**Comment by dreamorosi** [comment]
Hello, is there any chance to revisit this?

As mentioned above in the thread, the `cause` property is becoming more and more popular so having an idiomatic way of testing it would be really helpful.

**Comment by adrienZ** [comment]
This is how I do it 

```ts
import { it, expect } from "vitest";

    it("should test cause", async () => {
      const error = await myPromise().catch(e => JSON.parse(JSON.stringify(e)));
      expect(error).toStrictEqual({ cause: "myCause" });
  })
```

**Comment by hi-ogawa** [comment]
`expect.objectContaining({ cause: ... })` also allows assertions like:

https://github.com/vitest-dev/vitest/blob/257d49d61e1036c7bb54958f34efdf1ae382f03e/test/core/test/expect-poll.test.ts#L36-L46

**Comment by RobinTail** [comment]
Here is my approach — a custom serializer for `Error` only that takes `cause` into account:

```ts
// vitest.setup.ts
expect.addSnapshotSerializer({
  test: (subject) => subject instanceof Error,
  serialize: ({ name, message, cause }: Error) =>
    `[${name}: ${message}${cause ? ` (${cause})` : ""}]`,
});
```

Then you can do 

```ts
expect(someFunction).toThrowErrorMatchingSnapshot()
```
